Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp

```diff
@@ -29,118 +29,300 @@
-
-    auto root = makeUnique<TextureMapperLayerNode>(layerList.takeFirst());
-    buildTree(*root, layerList);
-
-    // Collect clip data
-    Vector<float> clipVertices;
-    traverseTree(*root, [&clipVertices](TextureMapperLayerNode& node) {
-        for (auto& polygon : node.polygons) {
-            auto toLayerTransform = polygon.layer->toSurfaceTransform().inverse();
-            if (polygon.isSplitted && toLayerTransform) {
-                polygon.clipVertexBufferOffset = clipVertices.size();
-
-                unsigned numVertices = polygon.geometry.numberOfVertices();
-                for (unsigned i = 0; i < numVertices; i++) {
-                    auto v = toLayerTransform->mapPoint(polygon.geometry.vertexAt(i));
-                    clipVertices.append(v.x());
-                    clipVertices.append(v.y());
-                }
+    Vector<Layer> layers;
+    for (size_t i = 0; i < textureMapperLayers.size(); ++i) {
+        auto layer = textureMapperLayers[i];
+        auto layerRect = layer->effectiveLayerRect();
+        auto transform = layer->toSurfaceTransform();
+        layers.append({ { layerRect, transform }, layer, false });
+    }
+
+    // Perform a broad-phase sweep-and-prune across all axes to identify potential intersections.
+    // By determining which layers might intersect, we can limit BSP cutting planes to those areas only,
+    // preventing unnecessary splitting of layers that are spatially distant.
+    // This optimization matters because the TextureMapper uses stencil operations to render split layers,
+    // and on some hardware, stencil usage is slow. Reducing unnecessary splits helps improve performance.
+    auto potentialIntersectionsXAxis = sweepAndPrune(layers, SweepAndPruneAxis::XAxis);
+    auto potentialIntersectionsYAxis = sweepAndPrune(layers, SweepAndPruneAxis::YAxis);
+    auto potentialIntersectionsZAxis = sweepAndPrune(layers, SweepAndPruneAxis::ZAxis);
+
+    // Filter intersections and take pairs that are present in all sweep and prune sets
+    SweepAndPrunePairs potentialIntersectionsXY;
+    std::set_intersection(
+        potentialIntersectionsXAxis.begin(), potentialIntersectionsXAxis.end(),
+        potentialIntersectionsYAxis.begin(), potentialIntersectionsYAxis.end(),
+        std::inserter(potentialIntersectionsXY, potentialIntersectionsXY.begin())
+    );
+    SweepAndPrunePairs potentialIntersections;
+    std::set_intersection(
+        potentialIntersectionsXY.begin(), potentialIntersectionsXY.end(),
+        potentialIntersectionsZAxis.begin(), potentialIntersectionsZAxis.end(),
+        std::inserter(potentialIntersections, potentialIntersections.begin())
+    );
+
+    // Determine if any layer pairs intersect on the ZY plane. An intersection implies that the rendering
+    // order is either ambiguous or overlapping, necessitating the use of a BSP tree for proper ordering.
```

Comment:
I think this comment is not correct. This is approximation.
For example, in the diagram below, layer A, B, C don't intersect.
However, in YZ plane, they intersect.

Reply:
Yes, they intersect in ZY plane and in that case BSP will handle it. This is on purpose, because it's very difficult to cover all non-intersecting scenarios without BSP. Some of non-intersecting cases need BSP because of ambiguous rendering order like in the test case attached to bug. If ZY plane intersection doesn't happen then we can be sure that  layers can be rendered without clip in z-order, otherwise BSP will handle it.
 In this scenario BSP doesn't do splitting either because split planes are coplanar to layers. BSP detects that layers are fully behind each other and renders them without clip.